### PR TITLE
Drop xterm for xfce terminal

### PR DESCRIPTION
**Pull request by AlanGriffiths**

### CODE REVIEW

**Review by townsend2010 — COMMENTED** [review]
Very cool!  Need to run `clang-format-12` on this:
```diff
--- src/platform/client/client_platform_linux.cpp	(before formatting)
+++ src/platform/client/client_platform_linux.cpp	(after formatting)
@@ -26,7 +26,8 @@
 {
     assert(!instance_name.isEmpty() && "Instance name cannot be empty");
     QProcess::startDetached(
-        "x-terminal-emulator", {"--initial-title", instance_name, "-e", QString("bash -c 'multipass shell %1 || read'").arg(instance_name)});
+        "x-terminal-emulator",
+        {"--initial-title", instance_name, "-e", QString("bash -c 'multipass shell %1 || read'").arg(instance_name)});
 }
```

**Comment by Saviq** [comment]
bors try

**Comment by Saviq** [comment]
Something's wrong with unicode, can anything be done about that?

photo_2021-08-21 19 34 02

**Comment by AlanGriffiths** [comment]
> Something's wrong with unicode, can anything be done about that?

I'd guess that it is a matter of adding a font that supplies the missing codepoints. Will experiment on Monday.

**Comment by AlanGriffiths** [comment]
Another problem (maybe just another manifestation):
```
$ locale -a
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_COLLATE to default locale: No such file or directory
C
C.UTF-8
POSIX
```

**Comment by AlanGriffiths** [comment]
> Something's wrong with unicode, can anything be done about that?

Just as a sanity check, this wasn't working right with xterm either:
```plain
ubuntu@test:~$ snap list
Name       Version   Rev    Tracking       Publisher   Notes
core18     20210722  2128   latest/stable  canonicalâ  base
core20     20210702  1081   latest/stable  canonicalâ  base
lxd        4.0.7     21029  4.0/stable/â¦   canonicalâ  -
multipass  1.7.1     5309   latest/stable  canonicalâ  -
snapd      2.51.3    12704  latest/stable  canonicalâ  snapd
```
Same example with xfce:
```
ubuntu@test:~$ snap list
Name       Version   Rev    Tracking       Publisher   Notes
core18     20210722  2128   latest/stable  canonical���  base
core20     20210702  1081   latest/stable  canonical���  base
lxd        4.0.7     21029  4.0/stable/���   canonical���  -
multipass  1.7.1     5309   latest/stable  canonical���  -
snapd      2.51.3    12704  latest/stable  canonical���  snapd
```

**Comment by AlanGriffiths** [comment]
OK, I finally noticed that I can set the encoding to Unicode [Terminal=>Set encoding=>Unicode] and that works. So, just need to figure out the right incantation.

**Comment by Saviq** [comment]
bors try

File: snap-wrappers/bin/launch-multipass-gui [hunk 1]
```diff
@@ -13,4 +13,41 @@ if [ -O "${real_wayland}" ]; then
   ln -sf "${real_wayland}.lock" "$XDG_RUNTIME_DIR"
 fi
 
+if [ ! -e "${XDG_CONFIG_HOME}/xfce4/terminal/terminalrc" ]; then
+  mkdir -p "${XDG_CONFIG_HOME}/xfce4/terminal/"
+  cat <<EOT > "${XDG_CONFIG_HOME}/xfce4/terminal/terminalrc"
+arw-file=
+console-provider=vt
```
Comment (Saviq) [inline snap-wrappers/bin/launch-multipass-gui:20]: Huh?

File: snap-wrappers/bin/launch-multipass-gui [hunk 2]
```diff
@@ -13,4 +13,41 @@ if [ -O "${real_wayland}" ]; then
...
+MiscHighlightUrls=TRUE
+MiscMiddleClickOpensUri=FALSE
+MiscCopyOnSelect=FALSE
+MiscShowRelaunchDialog=TRUE
+MiscRewrapOnResize=TRUE
+MiscUseShiftArrowsToScroll=FALSE
+MiscSlimTabs=FALSE
+MiscNewTabAdjacent=FALSE
+MiscSearchDialogOpacity=100
+MiscShowUnsafePasteDialog=TRUE
```
Comment (Saviq) [inline snap-wrappers/bin/launch-multipass-gui:49]: Should we leave these default?

File: snap-wrappers/bin/launch-multipass-gui [hunk 3]
```diff
@@ -3,4 +3,51 @@
...
 
+mkdir -p "$XDG_RUNTIME_DIR" -m 700
+
+real_xdg_runtime_dir=$(dirname "${XDG_RUNTIME_DIR}")
+real_wayland=${real_xdg_runtime_dir}/${WAYLAND_DISPLAY:-wayland-0}
+
+if [ -O "${real_wayland}" ]; then
+  ln -sf "${real_wayland}" "$XDG_RUNTIME_DIR"
+  ln -sf "${real_wayland}.lock" "$XDG_RUNTIME_DIR"
+fi
```
Comment (Saviq) [inline snap-wrappers/bin/launch-multipass-gui:14]: Do desktop helpers normally take care of that?

File: src/platform/client/client_platform_linux.cpp [hunk 4]
```diff
@@ -26,7 +26,8 @@ void mcp::open_multipass_shell(const QString& instance_name)
 {
     assert(!instance_name.isEmpty() && "Instance name cannot be empty");
     QProcess::startDetached(
-        "xterm", {"-title", instance_name, "-j", "-e", QString("multipass shell %1 || read").arg(instance_name)});
+        "x-terminal-emulator",
```
Comment (Saviq) [inline src/platform/client/client_platform_linux.cpp:29]: Going through `x-terminal-emulator` caters for running out of snap, does it?

**Review by Saviq — CHANGES_REQUESTED** [review]

File: snap-wrappers/bin/launch-multipass-gui (hunk 1 above)
Comment (AlanGriffiths) [reply snap-wrappers/bin/launch-multipass-gui:20]: Oops! benign copy-paste error - will fix.

File: snap-wrappers/bin/launch-multipass-gui (hunk 3 above)
Comment (AlanGriffiths) [reply snap-wrappers/bin/launch-multipass-gui:14]: Ah. Added while experimenting - will confirm if needed.

File: snap-wrappers/bin/launch-multipass-gui (hunk 2 above)
Comment (AlanGriffiths) [reply snap-wrappers/bin/launch-multipass-gui:49]: Suggestions welcome

File: src/platform/client/client_platform_linux.cpp (hunk 4 above)
Comment (AlanGriffiths) [reply src/platform/client/client_platform_linux.cpp:29]: No, but I didn't get around to deciphering what it does for this case. Dumping it in was a BFI approach to getting things working.

**Comment by AlanGriffiths** [comment]
bors try

File: snap/hooks/post-refresh [hunk 5]
```diff
@@ -1,3 +1,22 @@
...
+
+schema_dir="${SNAP}/usr/share/glib-2.0/schemas"
+if [ -n "$(ls -A $schema_dir/*.xml 2>/dev/null)" ]; then
+  ln -s $schema_dir/*.xml $GS_SCHEMA_DIR
+fi
+if [ -n "$(ls -A $schema_dir/*.override 2>/dev/null)" ]; then
+  ln -s $schema_dir/*.override $GS_SCHEMA_DIR
+fi
+
+$SNAP/usr/lib/${ARCH_TRIPLET}/glib-2.0/glib-compile-schemas "$GS_SCHEMA_DIR"
```
Comment (Saviq) [inline snap/hooks/post-refresh:22]: Why not call `install` here?

File: snap/hooks/post-refresh [hunk 6]
```diff
@@ -1,3 +1,22 @@
...
+mkdir -p "$GS_SCHEMA_DIR"
+rm -f "$GS_SCHEMA_DIR"/*
+
+schema_dir="${SNAP}/usr/share/glib-2.0/schemas"
+if [ -n "$(ls -A $schema_dir/*.xml 2>/dev/null)" ]; then
+  ln -s $schema_dir/*.xml $GS_SCHEMA_DIR
+fi
+if [ -n "$(ls -A $schema_dir/*.override 2>/dev/null)" ]; then
+  ln -s $schema_dir/*.override $GS_SCHEMA_DIR
+fi
```
Comment (Saviq) [inline snap/hooks/post-refresh:20]: Why is this different than in `install`?

File: snap-wrappers/bin/launch-multipass-gui [hunk 7]
```diff
@@ -3,4 +3,14 @@
 . client-common.sh
 link_autostart
 
+if [ ! -e "${XDG_CONFIG_HOME}/xfce4/terminal/terminalrc" ]; then
+  mkdir -p "${XDG_CONFIG_HOME}/xfce4/terminal/"
+  cat <<EOT > "${XDG_CONFIG_HOME}/xfce4/terminal/terminalrc"
+[Configuration]
+Encoding=UTF-8
+MiscMenubarDefault=FALSE
+ShortcutsNoMenukey=TRUE
```
Comment (Saviq) [inline snap-wrappers/bin/launch-multipass-gui:12]: Please add this:

```suggestion
ShortcutsNoMenukey=TRUE
FontUseSystem=TRUE
```

**Review by Saviq — CHANGES_REQUESTED** [review]
Works fine, handful of tweaks / questions inline.

File: snap-wrappers/bin/launch-multipass-gui (hunk 7 above)
Comment (AlanGriffiths) [reply snap-wrappers/bin/launch-multipass-gui:12]: Why? That looks horrible!

Screenshot from 2021-08-31 14-21-03

File: snap/hooks/post-refresh (hunk 5 above)
Comment (AlanGriffiths) [reply snap/hooks/post-refresh:22]: While that works now I didn't think the next person to change install would necessarily realise it was also invoked here.

Alternatively, why isn't install a link to post-refresh?

**Review by Saviq — APPROVED** [review]
Thanks!

bors merge

**Comment by Saviq** [comment]
bors retry

**Comment by AlanGriffiths** [comment]
bors retry

**Pull request merged**